samples:subsys:usb:webusb: esp32s3 after enabling spi ram, device configuration is not supported

### Describe the bug

Use esp32s3 chip.

https://github.com/zephyrproject-rtos/zephyr/pull/85600

### Regression

- [ ] This is a regression.

### Steps to reproduce

Edit `prj.conf` add `CONFIG_ESP_SPIRAM=y` and `CONFIG_SPIRAM_MODE_OCT=y`

`west build -b esp32s3_devkitm/esp32s3/procpu samples/subsys/usb/webusb`

Flash: 

`west flash`



### Relevant log output

**Before successful**:

```shell

ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x1 (POWERON),boot:0x2b (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fc8f5d0,len:0x22b8
load:0x40374000,len:0xb5c0
SHA-256 comparison failed:
Calculated: 28564cddd6ee50ea1e047a9cab985194199fa26f82a67a1435e35748202e229b
Expected: 0000000040270000000000000000000000000000000000000000000000000000
Attempting to boot anyway...
entry 0x4037a3d8
I (61) soc_init: ESP Simple boot
I (61) soc_init: compile time Sep 28 2025 23:18:02
W (62) soc_init: Unicore bootloader
I (62) soc_init: chip revision: v0.2
I (64) flash_init: Boot SPI Speed : 80MHz
I (68) flash_init: SPI Mode       : DIO
I (71) flash_init: SPI Flash Size : 8MB
I (75) boot: DRAM       : lma=00000020h vma=3fc8f5d0h size=022b8h (  8888)
I (81) boot: IRAM       : lma=000022e0h vma=40374000h size=0b5c0h ( 46528)
I (87) boot: IRAM       : lma=0000d8b8h vma=00000000h size=02740h ( 10048)
I (93) boot: IRAM       : lma=00010000h vma=42000000h size=0a848h ( 43080)
I (99) boot: IRAM       : lma=0001a850h vma=00000000h size=057a8h ( 22440)
I (105) boot: IRAM      : lma=00020000h vma=3c010000h size=02778h ( 10104)
I (112) boot: IROM      : lma=00010000h vma=42000000h size=0A846h ( 43078) map
I (118) boot: DROM      : lma=00020000h vma=3c010000h size=02780h ( 10112) map
I (136) boot: libc heap size 331 kB.
I (136) spi_flash: detected chip: gd
I (136) spi_flash: flash io: dio
W (136) spi_flash: Detected size(16384k) larger than the size in the binary image header(8192k). Using the size in the binary image header.
*** Booting Zephyr OS build v4.2.0-4553-g7fc8051b0f4b ***
[00:00:00.206,000] <wrn> udc_dwc2: Experimental DMA enabled
[00:00:00.248,000] <inf> main: USBD message: Bus reset
[00:00:00.426,000] <inf> main: USBD message: Bus reset
[00:00:00.502,000] <inf> sfunc: Configuration enabled
[00:00:00.502,000] <inf> main: USBD message: New device configuration
```

**After adding `CONFIG_ESP_SPIRAM=y` and `CONFIG_SPIRAM_MODE_OCT=y`**  :

```shell
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x1 (POWERON),boot:0x2b (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fc8ff20,len:0x28ec
load:0x40374000,len:0xbf10
SHA-256 comparison failed:
Calculated: fe65b42fd69e11643ba8811c334d17024cac06bce22163e25881dfd065e0cce1
Expected: 00000000c0170000000000000000000000000000000000000000000000000000
Attempting to boot anyway...
entry 0x4037a528
I (64) soc_init: ESP Simple boot
I (64) soc_init: compile time Sep 28 2025 22:51:39
W (64) soc_init: Unicore bootloader
I (64) soc_init: chip revision: v0.2
I (66) flash_init: Boot SPI Speed : 80MHz
I (70) flash_init: SPI Mode       : DIO
I (74) flash_init: SPI Flash Size : 8MB
I (77) boot: DRAM       : lma=00000020h vma=3fc8ff20h size=028ech ( 10476)
I (83) boot: IRAM       : lma=00002914h vma=40374000h size=0bf10h ( 48912)
I (89) boot: IRAM       : lma=0000e838h vma=00000000h size=017c0h (  6080)
I (95) boot: IRAM       : lma=00010000h vma=42000000h size=0c2c8h ( 49864)
I (101) boot: IRAM      : lma=0001c2d0h vma=00000000h size=03d28h ( 15656)
I (108) boot: IRAM      : lma=00020000h vma=3c010000h size=02c7ch ( 11388)
I (114) boot: IROM      : lma=00010000h vma=42000000h size=0C2C6h ( 49862) map
I (120) boot: DROM      : lma=00020000h vma=3c010000h size=02C80h ( 11392) map
I (138) boot: libc heap size 326 kB.
I (138) spi_flash: detected chip: gd
I (138) spi_flash: flash io: dio
W (139) spi_flash: Detected size(16384k) larger than the size in the binary image header(8192k). Using the size in the binary image header.
I (149) octal_psram: vendor id    : 0x0d (AP)
I (153) octal_psram: dev id       : 0x02 (generation 3)
I (158) octal_psram: density      : 0x03 (64 Mbit)
I (162) octal_psram: good-die     : 0x01 (Pass)
I (166) octal_psram: Latency      : 0x01 (Fixed)
I (171) octal_psram: VCC          : 0x01 (3V)
I (175) octal_psram: SRF          : 0x01 (Fast Refresh)
I (180) octal_psram: BurstType    : 0x01 (Hybrid Wrap)
I (185) octal_psram: BurstLen     : 0x01 (32 Byte)
I (190) octal_psram: Readlatency  : 0x02 (10 cycles@Fixed)
I (194) octal_psram: DriveStrength: 0x00 (1/1)
I (199) esp_psram: Found 8MB PSRAM device
I (202) esp_psram: Speed: 40MHz
I (936) esp_psram: SPI SRAM memory test OK
*** Booting Zephyr OS build v4.2.0-4553-g7fc8051b0f4b ***
[00:00:00.988,000] <wrn> udc_dwc2: Experimental DMA enabled
```

### Impact

Functional Limitation – Some features not working as expected, but system usable.

### Environment

OS: Debian
SHA: 7fc8051b0f4b4299034fb110f821adeba3445368

### Additional Context

_No response_

@sylvioalves 

I don't know which commit introduced it, but it must have been in the last few days

New Error:

Edit add: samples/subsys/usb/webusb/prj.conf 

```
CONFIG_ESP_SPIRAM=y
CONFIG_SPIRAM_MODE_OCT=y
```

```
west build -b esp32s3_devkitm/esp32s3/procpu samples/subsys/usb/webusb
[0/1] Re-running CMake...
Loading Zephyr default modules (Zephyr base (cached)).
-- Application: /home/life/Develop/platform/zephyr-project/zephyr/samples/subsys/usb/webusb
-- CMake version: 3.30.5
-- Cache files will be written to: /home/life/.cache/zephyr
-- Zephyr version: 4.2.99 (/home/life/Develop/platform/zephyr-project/zephyr)
-- Found west (found suitable version "1.4.0", minimum required is "0.14.0")
-- Board: esp32s3_devkitm, qualifiers: esp32s3/procpu
-- Found host-tools: zephyr 0.17.2 (/home/life/Develop/platform/zephyr-project/sdk)
-- Found toolchain: zephyr 0.17.2 (/home/life/Develop/platform/zephyr-project/sdk)
-- Found BOARD.dts: /home/life/Develop/platform/zephyr-project/zephyr/boards/espressif/esp32s3_devkitm/esp32s3_devkitm_procpu.dts
-- Generated zephyr.dts: /home/life/Develop/platform/zephyr-project/zephyr/build/zephyr/zephyr.dts
-- Generated pickled edt: /home/life/Develop/platform/zephyr-project/zephyr/build/zephyr/edt.pickle
-- Generated devicetree_generated.h: /home/life/Develop/platform/zephyr-project/zephyr/build/zephyr/include/generated/zephyr/devicetree_generated.h
Parsing /home/life/Develop/platform/zephyr-project/zephyr/samples/subsys/usb/webusb/Kconfig
Loaded configuration '/home/life/Develop/platform/zephyr-project/zephyr/boards/espressif/esp32s3_devkitm/esp32s3_devkitm_procpu_defconfig'
Merged configuration '/home/life/Develop/platform/zephyr-project/zephyr/samples/subsys/usb/webusb/prj.conf'
Configuration saved to '/home/life/Develop/platform/zephyr-project/zephyr/build/zephyr/.config'
Kconfig header saved to '/home/life/Develop/platform/zephyr-project/zephyr/build/zephyr/include/generated/zephyr/autoconf.h'
-- Espressif HAL path: /home/life/Develop/platform/zephyr-project/modules/hal/espressif
-- Image partition address: 0x0
-- Using ccache: /usr/bin/ccache
-- Found gen_kobject_list: /home/life/Develop/platform/zephyr-project/zephyr/scripts/build/gen_kobject_list.py
-- Configuring done (6.1s)
-- Generating done (0.2s)
-- Build files have been written to: /home/life/Develop/platform/zephyr-project/zephyr/build
[245/251] Linking C executable zephyr/zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /home/life/Develop/platform/zephyr-project/zephyr/build/zephyr/zephyr_pre0.map 
: && ccache /home/life/Develop/platform/zephyr-project/sdk/xtensa-espressif_esp32s3_zephyr-elf/bin/xtensa-espressif_esp32s3_zephyr-elf-gcc  -gdwarf-4 -Os zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/xtensa/core/offsets/offsets.c.obj  -T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map,/home/life/Develop/platform/zephyr-project/zephyr/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/xtensa/core/libarch__xtensa__core.a  zephyr/lib/libc/picolibc/liblib__libc__picolibc.a  zephyr/lib/libc/common/liblib__libc__common.a  zephyr/lib/net_buf/liblib__net_buf.a  zephyr/subsys/usb/device_next/libsubsys__usb__device_next.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/usb/udc/libdrivers__usb__udc.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/hwinfo/libdrivers__hwinfo.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L/home/life/Develop/platform/zephyr-project/zephyr/build/zephyr  zephyr/arch/common/libisr_tables.a  -fuse-ld=bfd  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn  -Wl,-no-pie  -specs=picolibc.specs  -DPICOLIBC_LONG_LONG_PRINTF_SCANF  -lgcc  -T/home/life/Develop/platform/zephyr-project/modules/hal/espressif/zephyr/esp32s3/src/linker/esp32s3.rom.alias.ld  -T/home/life/Develop/platform/zephyr-project/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.ld  -T/home/life/Develop/platform/zephyr-project/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.api.ld  -T/home/life/Develop/platform/zephyr-project/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.libgcc.ld  -T/home/life/Develop/platform/zephyr-project/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.newlib.ld  -T/home/life/Develop/platform/zephyr-project/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.version.ld  -T/home/life/Develop/platform/zephyr-project/modules/hal/espressif/zephyr/esp32s3/../../components/soc/esp32s3/ld/esp32s3.peripherals.ld  -Wl,--wrap=longjmp -L/home/life/Develop/platform/zephyr-project/sdk/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0 -lc -lgcc && cd /home/life/Develop/platform/zephyr-project/zephyr/build/zephyr && /home/linuxbrew/.linuxbrew/Cellar/cmake/3.30.5/bin/cmake -E true
/home/life/Develop/platform/zephyr-project/sdk/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/zephyr_pre0.elf section `.ext_ram.dummy' will not fit in region `ext_dram_seg'
/home/life/Develop/platform/zephyr-project/sdk/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: External SPIRAM overflowed.
/home/life/Develop/platform/zephyr-project/sdk/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: region `ext_dram_seg' overflowed by 1179648 bytes
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /home/linuxbrew/.linuxbrew/bin/cmake --build /home/life/Develop/platform/zephyr-project/zephyr/build
```

--- a/drivers/usb/udc/Kconfig.dwc2
+++ b/drivers/usb/udc/Kconfig.dwc2
@@ -37,6 +37,7 @@ config UDC_DWC2_PTI
 
 config UDC_DWC2_STACK_SIZE
 	int "UDC DWC2 driver internal thread stack size"
+	default 1024 if SOC_SERIES_ESP32S3
 	default 512
 	help
 	  DWC2 driver internal thread stack size.

--- a/drivers/usb/udc/udc_dwc2_vendor_quirks.h
+++ b/drivers/usb/udc/udc_dwc2_vendor_quirks.h
@@ -607,26 +607,38 @@ static inline int esp32_usb_otg_init(const struct device *dev,
 	return ret;
 }
 
+static inline int esp32_usb_otg_enable_clk(struct phy_context_t *phy_ctx)
+{
+	usb_wrap_ll_enable_bus_clock(true);
+
+	usb_wrap_ll_reset_register();
+	usb_wrap_hal_init(&phy_ctx->wrap_hal);
+
+#if USB_WRAP_LL_EXT_PHY_SUPPORTED
+	usb_wrap_hal_phy_set_external(&phy_ctx->wrap_hal, (phy_ctx->target == USB_PHY_TARGET_EXT));
+#endif
+
+	return 0;
+}
+
 static inline int esp32_usb_otg_enable_phy(struct phy_context_t *phy_ctx, bool enable)
 {
 	LOG_MODULE_DECLARE(udc_dwc2, CONFIG_UDC_DRIVER_LOG_LEVEL);
 
 	if (enable) {
-		usb_wrap_ll_enable_bus_clock(true);
-		usb_wrap_hal_init(&phy_ctx->wrap_hal);
-
-#if USB_WRAP_LL_EXT_PHY_SUPPORTED
-		usb_wrap_hal_phy_set_external(&phy_ctx->wrap_hal,
-					      (phy_ctx->target == USB_PHY_TARGET_EXT));
-#endif
-
+		usb_wrap_ll_phy_enable_pad(phy_ctx->wrap_hal.dev, true);
 		LOG_DBG("PHY enabled");
 	} else {
-		usb_wrap_ll_enable_bus_clock(false);
 		usb_wrap_ll_phy_enable_pad(phy_ctx->wrap_hal.dev, false);
-
 		LOG_DBG("PHY disabled");
 	}
+
+	return 0;
+}
+
+static inline int esp32_usb_otg_shutdown(struct phy_context_t *phy_ctx)
+{
+	usb_wrap_ll_enable_bus_clock(false);
 
 	return 0;
 }
@@ -660,6 +672,11 @@ static inline int esp32_usb_otg_enable_phy(struct phy_context_t *phy_ctx, bool e
 			&usb_otg_config_##n, &usb_otg_data_##n);		\
 	}									\
 										\
+	static int esp32_usb_otg_enable_clk_##n(const struct device *dev)	\
+	{									\
+		return esp32_usb_otg_enable_clk(&phy_ctx_##n);			\
+	}									\
+										\
 	static int esp32_usb_otg_enable_phy_##n(const struct device *dev)	\
 	{									\
 		return esp32_usb_otg_enable_phy(&phy_ctx_##n, true);		\
@@ -669,11 +686,18 @@ static inline int esp32_usb_otg_enable_phy(struct phy_context_t *phy_ctx, bool e
 	{									\
 		return esp32_usb_otg_enable_phy(&phy_ctx_##n, false);		\
 	}									\
+	static int esp32_usb_otg_shutdown_##n(const struct device *dev)		\
+	{									\
+		esp_intr_free(usb_otg_data_##n.int_handle);			\
+		return esp32_usb_otg_shutdown(&phy_ctx_##n);			\
+	}									\
 										\
 	const struct dwc2_vendor_quirks dwc2_vendor_quirks_##n = {		\
 		.init = esp32_usb_otg_init_##n,					\
+		.pre_enable = esp32_usb_otg_enable_clk_##n,			\
 		.post_enable = esp32_usb_otg_enable_phy_##n,			\
 		.disable = esp32_usb_otg_disable_phy_##n,			\
+		.shutdown = esp32_usb_otg_shutdown_##n,				\
 	};									\
 
 #define UDC_DWC2_IRQ_DT_INST_DEFINE(n)						\

